ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +200,99 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
 	return w.stack.PeekAll()
 }
 
-// pollNextBlock polls for the next block header to be added to the block stack.
-// If there are no blocks on the stack, it fetches the first block at the specified
-// `startBlockDepth` supplied at instantiation.
-func (w *Watcher) pollNextBlock() error {
-	var nextBlockNumber *big.Int
-	latestHeader, err := w.stack.Peek()
+// syncChain syncs our local state of the chain to the latest block found via
+// Ethereum RPC
+func (w *Watcher) syncChain() error {
+	latestHeader, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return err
 	}
-	if latestHeader == nil {
-		if w.startBlockDepth == rpc.LatestBlockNumber {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	latestBlockNumber := latestHeader.Number.Int64()
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
+	var lastStoredBlockNumber int64
+	if lastStoredHeader != nil {
+		lastStoredBlockNumber = lastStoredHeader.Number.Int64()
+	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestBlockNumber <= lastStoredBlockNumber {
+			return nil
+		}
+		numBlocksToFetch = int(latestBlockNumber - lastStoredBlockNumber)
 	}
-	if err != nil {
-		return err
+
+	if numBlocksToFetch >= constants.MaxBlocksStoredInNonArchiveNode {
+		return fmt.Errorf(
+			"cannot sync blocks more than or equal to %d blocks behind latest (%d). Your `BLOCK_POLLING_INTERVAL` is probably set too high",
+			constants.MaxBlocksStoredInNonArchiveNode,
+			numBlocksToFetch,
+		)
+	}
+
+	allEvents := []*Event{}
+	var syncErr error
```

Comment:
I like this solution better and it definitely seems less error-prone. Can you add a brief comment explaining what `syncErr` is, when it should be set, and when it will be returned? It is a little unusual and I think worth explicitly noting.